core/lib/rsa.c

```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/**
+ * @copyright 2018-2019 NXP
+ *
+ * @file    rsa.c
+ *
+ */
+/* Global includes */
+#include <crypto/crypto.h>
+#include <tee/tee_cryp_utl.h>
+#include <trace.h>
+#include <utee_defines.h>
+
+/**
+ * @brief   Allocate a RSA keypair
+ *
+ * @param[in]  key        Keypair
+ * @param[in]  size_bits  Key size in bits
+ *
+ * @retval TEE_SUCCESS                 Success
+ * @retval TEE_ERROR_BAD_PARAMETERS    Bad parameters
+ * @retval TEE_ERROR_OUT_OF_MEMORY     Out of memory
+ * @retval TEE_ERROR_NOT_IMPLEMENTED   Algorithm not implemented
+ */
+TEE_Result crypto_acipher_alloc_rsa_keypair(struct rsa_keypair *key,
+						size_t size_bits)
+{
+	/* Initialize all input key fields to 0 */
+	memset(key, 0, sizeof(*key));
+
+	/* Allocate the Public Exponent to maximum size */
+	key->e = crypto_bignum_allocate(256);
+	if (!key->e)
+		goto err_alloc_keypair;
+
+	/* Allocate the Private Exponent [d = 1/e mod LCM(p-1, q-1)] */
+	key->d = crypto_bignum_allocate(size_bits + (RSEC_OVERHEAD * 8));
```

Comment:
I'd prefer extending the bignum interface either with an extra parameter to `crypto_bignum_allocate()` indicating that space should be reserved for RSEC overhead. That way all bignum implementations will support RSEC directly instead of indirectly where we assume that a certain number of extra bits will need certain number of bytes.

Reply:
We can do that.

One thing i want to confirm here is When we are allocating the memory for key, we don't know if the key will be RSEC or PLAIN. We get to know only when we set or not set the TEE_EXTRACTABLE Flag.
So `crypto_acipher_alloc_rsa_keypair()` function we will always be allocating the bignums with required RSEC overhead.

Does above statement holds true ?

Reply:
That's OK, as long as it's only done in configurations which supports black keys.

Reply:
So `crypto_acipher_alloc_rsa_keypair()` will be like this ?

```
TEE_Result crypto_acipher_alloc_rsa_keypair(struct rsa_keypair *key,
						size_t size_bits)
{
	/* Initialize all input key fields to 0 */
	memset(key, 0, sizeof(*key));

	/* Allocate the Public Exponent to maximum size */
	key->e = crypto_bignum_allocate(256);
	if (!key->e)
		goto err_alloc_keypair;

	/* Allocate the Modulus (size_bits) [n = p * q] */
	key->n = crypto_bignum_allocate(size_bits);
	if (!key->n)
		goto err_alloc_keypair;

#ifdef CFG_RUNTIME_SEC
	/* Allocate the Private Exponent [d = 1/e mod LCM(p-1, q-1)] */
	key->d = crypto_bignum_allocate(size_bits + (RSEC_OVERHEAD * 8));
	if (!key->d)
		goto err_alloc_keypair;

	/* Allocate the prime number p of size (size_bits / 2) */
	key->p = crypto_bignum_allocate((size_bits + (RSEC_OVERHEAD * 8))/2);
	if (!key->p)
		goto err_alloc_keypair;

	/* Allocate the prime number q of size (size_bits / 2) */
	key->q = crypto_bignum_allocate((size_bits + (RSEC_OVERHEAD * 8))/2);
	if (!key->q)
		goto err_alloc_keypair;

	/* Allocate dp (size_bits / 2) [d mod (p-1)] */
	key->dp = crypto_bignum_allocate((size_bits + (RSEC_OVERHEAD * 8))/2);
	if (!key->dp)
		goto err_alloc_keypair;

	/* Allocate dq (size_bits / 2) [d mod (q-1)] */
	key->dq = crypto_bignum_allocate((size_bits + (RSEC_OVERHEAD * 8))/2);
	if (!key->dq)
		goto err_alloc_keypair;

	/* Allocate qp (size_bits / 2) [1/q mod p] */
	key->qp = crypto_bignum_allocate((size_bits + (RSEC_OVERHEAD * 8))/2);
	if (!key->qp)
		goto err_alloc_keypair;
#else 
	/* Allocate the Private Exponent [d = 1/e mod LCM(p-1, q-1)] */
	key->d = crypto_bignum_allocate(size_bits);
	if (!key->d)
		goto err_alloc_keypair;

	/* Allocate the prime number p of size (size_bits / 2) */
	key->p = crypto_bignum_allocate(size_bits/2);
	if (!key->p)
		goto err_alloc_keypair;

	/* Allocate the prime number q of size (size_bits / 2) */
	key->q = crypto_bignum_allocate(size_bits);
	if (!key->q)
		goto err_alloc_keypair;

	/* Allocate dp (size_bits / 2) [d mod (p-1)] */
	key->dp = crypto_bignum_allocate(size_bits/2);
	if (!key->dp)
		goto err_alloc_keypair;

	/* Allocate dq (size_bits / 2) [d mod (q-1)] */
	key->dq = crypto_bignum_allocate(size_bits/2);
	if (!key->dq)
		goto err_alloc_keypair;

	/* Allocate qp (size_bits / 2) [1/q mod p] */
	key->qp = crypto_bignum_allocate(size_bits/2);
	if (!key->qp)
		goto err_alloc_keypair;
#endif 
	return TEE_SUCCESS;

err_alloc_keypair:
	crypto_bignum_free(key->e);
	crypto_bignum_free(key->d);
	crypto_bignum_free(key->n);
	crypto_bignum_free(key->p);
	crypto_bignum_free(key->q);
	crypto_bignum_free(key->dp);
	crypto_bignum_free(key->dq);
	crypto_bignum_free(key->qp);

	return TEE_ERROR_OUT_OF_MEMORY;
}


```

Reply:
Along that way, I'm sure we can get rid of the big ugly ifdef though.

Reply:
Yeah sure 👍